File: java/secret-handshake/src/main/java/HandshakeCalculator.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;

class HandshakeCalculator {
    private List<Signal> handshake = new ArrayList<>();

    List<Signal> calculateHandshake(int number) {
        List<String> reversedBinary = getReversedBinaryArray(number);

        for (int i = 0; i < reversedBinary.size(); i++) {
            if (reversedBinary.get(i).equals("1")) {
                addSignal(Signal.find(i));
            }
        }

        if (number > 15) Collections.reverse(handshake);

        return handshake;
    }

    private List<String> getReversedBinaryArray(int number) {
        List<String> chars = Arrays.asList(Integer.toBinaryString(number).split(""));
        Collections.reverse(chars);
        return chars;
    }

    private void addSignal(Signal signal) {
        if (signal != null) handshake.add(signal);
    }
}

enum Signal {
    WINK(0), DOUBLE_BLINK(1), CLOSE_YOUR_EYES(2), JUMP(3);

    private int value;

    Signal(int value) {
        this.value = value;
    }

    static Signal find(int value) {
        Signal[] signals = values();

        for (Signal signal : signals) {
            if (signal.value == value) {
                return signal;
            }
        }

        return null;
    }
}

File: kotlin/secret-handshake/src/main/kotlin/HandshakeCalculator.kt
class HandshakeCalculator {
    companion object {
        fun calculateHandshake(value: Int): List<Signal> {
            val binaryReversed = Integer.toBinaryString(value).reversed()
            val signalList = binaryReversed.mapIndexedNotNull { index, char -> getSignal(char, index) }
            return if (value <= 15) signalList else signalList.reversed()
        }

        private fun getSignal(char: Char, index: Int): Signal? {
            return when (char) {
                '1' -> Signal.values().find { it.value == index }
                else -> null
            }
        }
    }

    enum class Signal(val value: Int) {
        WINK(0), DOUBLE_BLINK(1), CLOSE_YOUR_EYES(2), JUMP(3)
    }
}